packages/indexer/src/indexers/amm-market/amm-market-account-indexer.ts

```diff
@@ -0,0 +1,69 @@
+import { AccountInfoIndexer } from "../account-info-indexer";
+import { rpcReadClient } from "../../connection";
+import { AccountInfo, Context, PublicKey } from "@solana/web3.js";
+import { schema, usingDb } from "@metadaoproject/indexer-db";
+import { Err, Ok } from "../../match";
+import { BN } from "@coral-xyz/anchor";
+
+type TwapRecord = typeof schema.twaps._.inferInsert;
+
+export enum AmmAccountIndexerError {
+  GeneralError = "GeneralError",
+}
+
+// Doing this rather than class implements pattern due to
+// https://github.com/microsoft/TypeScript/issues/41399
+export const AmmMarketAccountUpdateIndexer: AccountInfoIndexer = {
+  index: async (
+    accountInfo: AccountInfo<Buffer>,
+    account: PublicKey,
+    context?: Context
+  ) => {
+    try {
+      const ammMarketAccount = await rpcReadClient.markets.amm.decodeMarket(
+        accountInfo
+      );
+      //index twap
+
+      // agg is 0 so skipping
+      if (ammMarketAccount.oracle.aggregator.toNumber() === 0)
+        return Ok({ acct: account.toString() });
+
+      const twapCalculation: BN = ammMarketAccount.oracle.aggregator.div(
+        ammMarketAccount.oracle.lastUpdatedSlot.sub(
+          ammMarketAccount.createdAtSlot
+        )
+      );
+      const twapNumber: number = twapCalculation.toNumber();
+      const newTwap: TwapRecord = {
+        curTwap: BigInt(twapNumber),
+        marketAcct: account.toString(),
+        observationAgg: ammMarketAccount.oracle.aggregator
+          .toNumber()
+          .toString(),
+        proposalAcct: ammMarketAccount.proposal.toString(),
+        // alternatively, we could pass in the context of the update here
+        updatedSlot: context
+          ? BigInt(context.slot)
+          : BigInt(ammMarketAccount.oracle.lastUpdatedSlot.toNumber()),
+        createdAt: new Date(),
+      };
+
+      const twapUpsertResult = await usingDb((db) =>
+        db
+          .insert(schema.twaps)
```

Comment:
The other thing to consider here is to construct a series of inserts and then commit them upon completion, to save on DB updates / writes.

Reply:
ooh nice optimization for sure

Reply:
but I guess in this case we are just reacting to a single account update, so no guarantee it's in like a series or anything like that. So for this I think doing one insert still likely makes the most sense.

Reply:
okay after discussion, we will need to consider this as a performance concern going forward.